src/Waives.Http/RequestHandling/AccessTokenService.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Memory;
+using Polly;
+using Polly.Caching;
+using Polly.Caching.Memory;
+using Waives.Http.Responses;
+
+namespace Waives.Http.RequestHandling
+{
+    internal class AccessTokenService : IDisposable
+    {
+        private readonly string _clientId;
+        private readonly string _clientSecret;
+        private readonly IHttpRequestSender _requestSender;
+        private readonly MemoryCache _cache;
+        private readonly AsyncCachePolicy<AccessToken> _cachePolicy;
+
+        internal AccessTokenService(string clientId, string clientSecret, IHttpRequestSender requestSender)
+        {
+            _clientId = clientId;
+            _clientSecret = clientSecret;
+            _requestSender = requestSender;
+
+            _cache = new MemoryCache(new MemoryCacheOptions());
+            _cachePolicy = Policy.CacheAsync(
+                new MemoryCacheProvider(_cache).AsyncFor<AccessToken>(),
+                new ResultTtl<AccessToken>(t => new Ttl(t.LifeTime)),
+                (_, __, ___) => { });
```

Comment:
`onCacheError` is currently a no-op. I'm not sure what the desired behaviour is if an error occurs: retries will have already taken place, and the result may not be a valid `AccessToken`.

Reply:
throw an explicit exception explaining that we unable get a valid token? I think desired behaviour is "throws an exception that makes it clear what happened". Otherwise we'll get an error relating to a side-effect/knock-on of the "main" request failing with missing token I think?

Reply:
Looking at [the Polly docs](https://github.com/App-vNext/Polly/wiki/Cache#oncacheerror), this is a kind of event handler, so throwing an exception isn't appropriate:

> ### onCacheError
> 
> An optional `onCacheError` delegate allows specific code to be executed (for example for logging), if any call to the underlying `cacheProvider` throws an exception. If the `onCacheError` delegate is configured, it is used for both `onCacheGetError` and `onCachePutError`.

We could write a log message, though.

Reply:
We now log an error message, as of 86538c2.